### ISSUE TITLE
yt-dlp not properly internally handling unicode in titles.

### ISSUE THREAD

**Issue by RichyT** [yt-dlp/yt-dlp#5341]
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting a bug unrelated to a specific site
- [X] I've verified that I'm running yt-dlp version **2022.10.04** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/yt-dlp/yt-dlp/wiki/FAQ#video-url-contains-an-ampersand--and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Provide a description that is worded well enough to be understood

yt-dlp not properly internally handling unicode in titles.

Trying to download audio from a video. There is a unicode character in the title and yt-dlp errors out on this character. If the error was related to trying to open a file with an invalid character and hitting a filesystem limitation, I would not consider this a bug but it seems that it is happening internal to yt-dlp so I figure it is.

Running with --restrict-filenames bypasses this issue.

If this is expected behavior and marked as a wontfix, that's fine (as a developer, I would attempt to fix it. Though my unicode experience is weak). This may help someone else running into the same issue.

Video is: https://www.youtube.com/watch?v=45Tqi-KYzDE

ERROR: 'latin-1' codec can't encode character '\uff1a' in position 24: ordinal not in range(256)

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
$ yt-dlp https://www.youtube.com/watch?v=45Tqi-KYzDE -f 139 -vU                 
[debug] Command-line config: ['https://www.youtube.com/watch?v=45Tqi-KYzDE', '-f', '139', '-vU']
[debug] Encodings: locale ISO-8859-1, fs iso8859-1, pref ISO-8859-1, out ISO-8859-1, error ISO-8859-1, screen ISO-8859-1
[debug] yt-dlp version 2022.10.04 [4e0511f] (zip)
[debug] Python 3.7.2 (CPython 64bit) - Linux-3.10.17-x86_64-AMD_A6-5400K_APU_with_Radeon-tm-_HD_Graphics-with-slackware-14.1 (glibc 2.2.5)
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg 3.2.4, ffprobe 3.2.4, rtmpdump 2.4
[debug] Optional libraries: sqlite3-2.6.0
[debug] Proxy map: {}
[debug] Loaded 1690 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Latest version: 2022.10.04, Current version: 2022.10.04
yt-dlp is up to date (2022.10.04)
[debug] [youtube] Extracting URL: https://www.youtube.com/watch?v=45Tqi-KYzDE
[youtube] 45Tqi-KYzDE: Downloading webpage
[youtube] 45Tqi-KYzDE: Downloading android player API JSON
[youtube] 45Tqi-KYzDE: Downloading MPD manifest
[youtube] 45Tqi-KYzDE: Downloading MPD manifest
[debug] Sort order given by extractor: quality, res, fps, hdr:12, source, vcodec:vp9.2, channels, acodec, lang, proto
[debug] Formats sorted by: hasvid, ie_pref, quality, res, fps, hdr:12(7), source, vcodec:vp9.2(10), channels, acodec, lang, proto, filesize, fs_approx, tbr, vbr, abr, asr, vext, aext, hasaud, id
[info] 45Tqi-KYzDE: Downloading 1 format(s): 139
ERROR: 'latin-1' codec can't encode character '\uff1a' in position 24: ordinal not in range(256)
Traceback (most recent call last):
  File "/usr/local/bin/yt-dlp/yt_dlp/YoutubeDL.py", line 1477, in wrapper
    return func(self, *args, **kwargs)
  File "/usr/local/bin/yt-dlp/yt_dlp/YoutubeDL.py", line 1574, in __extract_info
    return self.process_ie_result(ie_result, download, extra_info)
  File "/usr/local/bin/yt-dlp/yt_dlp/YoutubeDL.py", line 1632, in process_ie_result
    ie_result = self.process_video_result(ie_result, download=download)
  File "/usr/local/bin/yt-dlp/yt_dlp/YoutubeDL.py", line 2733, in process_video_result
    self.process_info(new_info)
  File "/usr/local/bin/yt-dlp/yt_dlp/YoutubeDL.py", line 3192, in process_info
    dl_filename = existing_video_file(full_filename, temp_filename)
  File "/usr/local/bin/yt-dlp/yt_dlp/YoutubeDL.py", line 3087, in existing_video_file
    default_overwrite=False)
  File "/usr/local/bin/yt-dlp/yt_dlp/YoutubeDL.py", line 2923, in existing_file
    existing_files = list(filter(os.path.exists, orderedSet(filepaths)))
  File "/usr/lib64/python3.7/genericpath.py", line 19, in exists
    os.stat(path)
UnicodeEncodeError: 'latin-1' codec can't encode character '\uff1a' in position 24: ordinal not in range(256)
```

**Comment by RichyT** [comment]
Looking at the debug, it's possible that this is YouTube specific but I suspect that it is not.

**Comment by Grub4K** [comment]
What filesystem are you using? Does it support the default extended replacement character for `:`, `：`? The call to `os.stat` fails because it tries to encode the filename into `latin-1` using that replacement character.

Maybe replacements for filenames should respect the filesystem encoding?

**Comment by RichyT** [comment]
The filesystem is ext4. Hard to tell what characters you're using there but the ASCII ':' works fine and is translated properly if the --restrict-filenames option is used.

I have no idea why the colon is encoded that way but I know Scott uses speech recognition when creating his episode titles and I believe he uses Apple devices.

**Comment by Grub4K** [comment]
The second character presented it `'\uff1a'`.
`:` in filenames are getting replaced by `\uff1a` unless `--restrict-filenames` is provided if I understand correctly.

Is `LC_CTYPE` set and, if so, what value does it hold? Explicitly setting it to `en_US.UTF-8` could work in that case as a workaround.

I would try `touch '1：1'` from within shell and `open('2\uff1a2')` (maybe `os.stat('2\uff1a2')` as well) from within python and see if those are working as expected before and after setting the variable.

**Comment by RichyT** [comment]
LC_CTYPE appears to be set to en_US. Setting it like you suggested allowed the download. 

I think cutting and pasting your touch command just gave me a regular colon. Not sure how to enter a unicode version. I'll have to wait on trying the python commands.

**Comment by RichyT** [comment]
Given that "restricted-filenames" is kind of a 'safe mode' for filenames, perhaps the expected behavior here would be that yt-dlp should set its locale to that of the retrieved data before using it? Or potentially, restricted-filenames (or something similar) should be the default in some circumstances?

**Comment by Grub4K** [comment]
The decoding process from HTTP/HTML has nothing to do with the way it is encoding that name for the fs. Encoding information has to be built separately or what am I misunderstanding?

Since your environment doesn't seem to yield the correct encoding it fails to encode.

**Comment by RichyT** [comment]
I'd need to poke in deeper but typically, when the filename (based on the title) is retrieved from YouTube, the page will come with information saying how it is encoded. yt-dlp now hold that title but when it passes it to the os calls, it uses my local system encoding which is why things go wrong.

The thing is, yt-dlp should know the encoding if youtube is doing their thing correctly so should be able to set things so that the os calls can interpret the filename correctly. *OR* yt-dlp should translate the filename from the provided encoding to the local system encoding. Either is reasonable and either may be desirable for different reasons. The former can be accomplished manually by me changing my LC_CTYPE (permanently changing this on my environment is *not* a reasonable option) and the latter is approximated (but not properly) by using restrict-filename.

**Comment by Grub4K** [comment]
Since I cannot recreate your environment, I sadly cannot help much.
If you create a file `1：1` (notice this is "Fullwidth Colon") through the neccesary means, what are the results of `os.stat('1\uff1a1'.encode('utf8'))` and `os.stat('1\uff1a1')` **without** setting `LC_CTYPE`?

If you can figure out how to successfully call `os.stat` that way this could get implemented, which might be plausible given that `stat(2)` takes a `char*`. Perhaps it is better treating this as a Python problem first.

Potentially relevant links:
- [Python docs](https://docs.python.org/3/library/os.html#os.stat)
- [CPython Implementation](https://github.com/python/cpython/blob/main/Modules/posixmodule.c#L2456)
- [stat(2) man](https://linux.die.net/man/2/stat)

**Comment by RichyT** [comment]
I'm going to be busy most of today so I'll get to those tests later since I'm not used to Python programming (sure it won't take long). I was able to get the file created when I set the LC_TYPE, as you suggested but when I view it with LC_CTYPE as the default, it has weird characters in it. This is why I now think that the translation should probably be the default.

What I think this really is is that the way encoding are handled on the filesystem are a bit of a hack (though understandably so). I'll try and provide some code samples of what I'm talking about later.

**Comment by pukkandan** [comment]
I think you are completely misunderstanding the issue. Python handles all filenames as utf-8. If your system does not allow unicode, either due to filesystem limitations or due to your configuration, you have to use `--restrict-filename` (or `--replace-in-metadata` for more fine-tuned control). There is nothing we can do about it - it's not yt-dlp's responsibility to fix/override your system configuration

The specific issue you've reported can also be worked around with `--compat-options filename-sanitization` , but that's not a general solution - it'll still break for other unicode characters

Related: https://github.com/yt-dlp/yt-dlp/issues/4547

**Comment by pukkandan** [comment]
> There is nothing we can do about it - it's not yt-dlp's responsibility to fix/override your system configuration

Actually, we can probably sanitize the filename to the correct `filesystem_encoding()`. Let me see whether that's feasible

**Comment by Lesmiscore** [comment]
Imagine a "convenient" encoding like that just forbids `\/:*?"<>|` - I've never heard of it
I guess it should be relied on the OS rather than `filesystem_encoding()`. Of course both should be taken care of

**Comment by RichyT** [comment]
>There is nothing we can do about it - it's not yt-dlp's responsibility to fix/override your system configuration

Let me rephrase that for you. The user requests yt-dlp to download a video, yt-dlp makes assumptions about the filename to be used, attempts to use it then errors out when the filename turns out to be invalid.

I mean, I'm fine with the workaround. My interest here is simply from good coding practices.

**Comment by Grub4K** [comment]
> Actually, we can probably sanitize the filename to the correct `filesystem_encoding()`.

That would be what you want if I understood you correctly.

The fix would affect users with incorrect setup which would need to use the `LC_CTYPE` workaround afterwards. This should be fine as it would have been an incorrect setup to begin with.

**Comment by RichyT** [comment]
Both of the os.stat commands requested by grub4k give file-not-found errors. I am not sure I have generated the filename correctly. I downloaded the video with the LC_CTYPE then renamed the file with filename completion and removing extraneous text. The bytecodes of the filename (obtained by piping ls into od-c) are  357 274 232 in both files.

**Comment by RichyT** [comment]
>os.stat('1\uff1a1'.encode('utf8')) and os.stat('1\uff1a1') without setting LC_CTYPE?

I think something is up with the escaping here. Both of these result if file-not-found. When I tried
os.open('1\uff1a1'.encode('utf8'),os.O_WRONLY|os.O_CREAT);
os.open('1\uff1a1',os.O_WRONLY|os.O_CREAT);

both resulted in a file which literally has the name '1\uff1a1'.

**Comment by RichyT** [comment]
Hey guys, I checked and the latest version of Slackware has moved to en_US.utf8. So while I still think it could be handled a bit better, it seems likely that you won't be seeing this issue too often. I wouldn't blame you for not wanting to spend much time on this but might be worth bearing in mind if it's ever an issue with other languages or something (though I'm not sure how likely that is with utf8?).

**Comment by Grub4K** [comment]
> both resulted in a file which literally has the name '1\uff1a1'.

That is unfortunate. Maybe we cannot write a bytes names through `os`? If we cannot, then I think this cannot be fixed by yt-dlp directly and has to be moved to Python core (if its even a valid concern).

**Comment by RichyT** [comment]
I think I know what you're trying to get at. Unfortunately, I can't devote much time to this in the immediate future but when I can, I'll see if I can get it working.

**Comment by milahu** [comment]
`--replace-in-metadata` does not work to replace unicode chars

error: `[MetadataParser] Did not find 'x' in title`

```sh
#! /usr/bin/env bash

a=()

a+=(yt-dlp)
a+=(--ignore-config)
a+=("https://www.youtube.com/watch?v=Cmfm20hsozI")

# https://www.fileformat.info/info/unicode/char/ff5c/index.htm # FULLWIDTH VERTICAL LINE

# [MetadataParser] Did not find '\\xEF\\xBD\\x9C' in title
#a+=(--replace-in-metadata title '\xEF\xBD\x9C' '|')

# [MetadataParser] Did not find '｜' in title
#a+=(--replace-in-metadata title $'\xEF\xBD\x9C' '|') # UTF-8

# yt-dlp: error: --replace-in-metadata title '\udcff\' '|' is invalid; bad escape (end of pattern) at position 1
#a+=(--replace-in-metadata title $'\xFF\x5C' '|') # UTF-16

# [MetadataParser] Did not find '｜' in title
a+=(--replace-in-metadata title $'\uFF5C' '|') # UTF-16

exec "${a[@]}"
```

```console
$ echo -n *.mkv | hexdump -C
00000000  4d 43 20 42 4f 47 59 20  2d 20 46 4c 55 47 4d 4f  |MC BOGY - FLUGMO|
00000010  44 55 53 20 28 70 72 6f  64 2e 20 4f 6e 6b 65 6c  |DUS (prod. Onkel|
00000020  2d 42 29 20 ef bd 9c 20  4e 4d 4b 20 45 4e 54 45  |-B) ... NMK ENTE|
00000030  52 54 41 49 4e 4d 45 4e  54 2d 43 6d 66 6d 32 30  |RTAINMENT-Cmfm20|
00000040  68 73 6f 7a 49 2e 6d 6b  76                       |hsozI.mkv|

$ echo -n $'\xEF\xBD\x9C' | hexdump -C
00000000  ef bd 9c                                          |...|
```

**Comment by milahu** [comment]
> This error doesn't tell anything on why it occured

this should not be an error in the first place, only a warning

yt-dlp spends minutes on downloading megabytes of video streams
and at 99.9% progress it throws this stupid error and deletes all files?
sounds like boomer technology, where resources have no cost

yt-dlp should replace bad chars with `-` by default, which works for all filesystems
and of course, expose an option, something like `--replace-bad-chars-with`
which can be an empty string to ignore bad chars

to restore the original name, there is `--write-info-json`

[make python replace un-encodable chars with a string by default](https://stackoverflow.com/questions/1933184/make-python-replace-un-encodable-chars-with-a-string-by-default)

```
#!/usr/bin/env python3

import codecs

def some_handler(exception):
    return (b"-", exception.end)

codecs.register_error("some_handler", some_handler)

s = 'A\uff1aB'

_bytes = s.encode("latin1", errors="ignore") # b'AB'
_bytes = s.encode("latin1", errors="replace") # b'A?B'
_bytes = s.encode("latin1", errors="some_handler") # b'A-B'

print(repr(_bytes))
```

**Comment by pukkandan** [comment]
https://github.com/yt-dlp/yt-dlp/issues/5341#issuecomment-1289098802

PR is welcome

**Comment by pukkandan** [comment]
Related: https://github.com/yt-dlp/yt-dlp/pull/3561

**Comment by dirkf** [comment]
>I don't fully understand. Could you explain?

Problem 1 is to pass a proposed filename (pathname) through the Python API. My understanding is that you have to respect `sys.getfilesystemencoding()` at that stage. By now, [this may be a historical problem](https://peps.python.org/pep-0529/) but OP's filesystem encoding was POSIX latin-1 rather than utf-8, albeit with a slightly antique Linux[^1]. Upstream made `encodeFilename()` a no-op for Py3 in fit of optimism rather round-tripping with the filesystem encoding as in #3561.

Then the O/S routines called from the Python library receive this filename and pass it to the appropriate filesystem driver. The driver should handle any mismatch between the O/S routines and the on-disk directory encoding: testing needed.

Even if that "works" Problem 2 is that the actual filesystem may not support the same characters as the Python API, say if the system is a Unix-like O/S with a FAT32 mounted drive ([practical example, template `%(title)s>%(id)s.%(ext)s`](https://old.reddit.com/r/linuxquestions/comments/pgg20f).

Sanitising the proposed filename for the known problem Windows filesystem characters may be (has been?) adequate in practice for that issue, since there's no sign of novel filesystems that don't accept POSIX filenames.

In comparison, the long file name problem has more contributory factors so making solutions that don't involve an actual test of the filename unreliable.

[^1] Up-to-date compared with 2.6.18 ROM that yt-dl is running on.
